internal/access/destinationcredential.go

```diff
@@ -0,0 +1,189 @@
+package access
+
+import (
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/gin-gonic/gin"
+
+	"github.com/infrahq/infra/api"
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/uid"
+)
+
+type ListDestinationCredentialResponse struct {
+	Items          []models.DestinationCredential
+	MaxUpdateIndex int64
+}
+
+func (l ListDestinationCredentialResponse) UpdateIndex() int64 {
+	return l.MaxUpdateIndex
+}
+
+func (l ListDestinationCredentialResponse) ItemCount() int {
+	return len(l.Items)
+}
+
+func CreateDestinationCredential(c *gin.Context, destination string) (*models.DestinationCredential, error) {
+	rCtx := GetRequestContext(c)
+	// does the user have a grant to this destination?
+	grants, err := data.ListGrants(rCtx.DBTxn, data.ListGrantsOptions{
+		ByDestination:              destination,
+		BySubject:                  uid.NewIdentityPolymorphicID(rCtx.Authenticated.User.ID),
+		IncludeInheritedFromGroups: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("ListGrants: %w", err)
+	}
+	if len(grants) == 0 {
+		return nil, internal.ErrUnauthorized
+	}
+
+	dest, err := data.GetDestination(rCtx.DBTxn, data.GetDestinationOptions{ByName: destination})
+	if err != nil {
+		return nil, fmt.Errorf("GetDestination: %w", err)
+	}
+
+	cr := &models.DestinationCredential{
+		ID:                 uid.New(),
+		OrganizationMember: models.OrganizationMember{OrganizationID: rCtx.Authenticated.User.OrganizationID},
+		RequestExpiresAt:   time.Now().Add(2 * time.Minute),
+		UserID:             rCtx.Authenticated.User.ID,
+		DestinationID:      dest.ID,
+	}
+	err = data.CreateDestinationCredential(rCtx.DBTxn, cr)
+	if err != nil {
+		return nil, fmt.Errorf("CreateDestinationCredential: %w", err)
+	}
+	logging.Debugf("Creating DestinationCredential with id %d, orgID %d", cr.ID, cr.OrganizationID)
+
+	// commit the transaction so others can see the request
+	err = rCtx.DBTxn.Commit()
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for the request to be filled; listen for the notify on the specific record req.ID
+	listener, err := data.ListenForNotify(c, rCtx.DataDB, data.ListenForNotifyOptions{
```

Comment:
This is my primary concern. This approach is what ends up requiring `SafeCommit`.

I think it's unexpected that a `POST` request blocks for up to 5 minutes. 

Bruce made this suggestion on the previous draft PR. I think we should split this up into two API calls. A short lived `POST` that submits the request, and then a long polling `GET` to receive the response. 

I think that approach would also make the API request/response structs easier to follow, without the need for big block comments about how some subset of the fields will be used.

There's another problem hidden here, which is that we wait up to 5 minutes for the write to happen, where as normally we timeout write requests at 1 minute. We could fix that by setting a shorter deadline on the context used for the first part of this request, but the separate endpoints seems like it fixes more problems.

Is there anything that becomes more difficult with separate API endpoints for this?

Does this operation not fit into the `access.blockingRequest` helper that we extracted in this PR? Would it if we split the endpoint into two?

Reply:
Ignoring the fact that under normal conditions this flow takes 300ms, why would it be unexpected that a POST blocks for 5 minutes but it's normal for a GET request? There's nothing in the http spec that would indicate a POST should respond quickly. 

The block comments are required anyway, because they come from the db table. You can split up the request/response structs currently without changing which endpoint returns with the results, so the two issues are not tied together. 

Adjusting the timeouts to match is fine, 5 minutes is not a requirement anywhere.

It doesn't fit into the existing `blockingRequest` structure because it does not use an updateIndex, it simply wants to know when the pg_notification happens.